### ISSUE TITLE
ExoPlayer filters odd (i.e. not even) dimension video formats

### ISSUE THREAD

**Issue by RikHeijdens** [google/ExoPlayer#1071]
After upgrading to ExoPlayer 1.5.3 (from 1.5.0) we found out that ExoPlayer started reporting an incorrect amount of tracks of the following mpd: 

```
<MPD xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="urn:mpeg:dash:schema:mpd:2011" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 http://standards.iso.org/ittf/PubliclyAvailableStandards/MPEG-DASH_schema_files/DASH-MPD.xsd" type="static" mediaPresentationDuration="PT0H14M48.00S" timeShiftBufferDepth="PT1S" minimumUpdatePeriod="PT1H" maxSegmentDuration="PT3S" minBufferTime="PT1S" profiles="urn:mpeg:dash:profile:isoff-live:2011,urn:com:dashif:dash264">
  <Period id="1" start="PT0S">
    <AdaptationSet group="1" lang="en_stereo" mimeType="audio/mp4" minBandwidth="130358" maxBandwidth="130358" segmentAlignment="true">
      <Representation id="128kbps en" bandwidth="130358" codecs="mp4a.40.2" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <SegmentTemplate timescale="48000" duration="95232" media="audio/stereo/en/128kbit/segment_$Number$.m4s" initialization="audio/stereo/en/128kbit/init.mp4" startNumber="1"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="2" lang="no-voices_stereo" mimeType="audio/mp4" minBandwidth="130395" maxBandwidth="130395" segmentAlignment="true">
      <Representation id="128kbps karaoke" bandwidth="130395" codecs="mp4a.40.2" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <SegmentTemplate timescale="48000" duration="95232" media="audio/stereo/none/128kbit/segment_$Number$.m4s" initialization="audio/stereo/none/128kbit/init.mp4" startNumber="1"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="3" lang="en_surround" mimeType="audio/mp4" minBandwidth="321836" maxBandwidth="321836" segmentAlignment="true">
      <Representation id="320kbps en" bandwidth="321836" codecs="mp4a.40.5" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="6"/>
        <SegmentTemplate timescale="48000" duration="95232" media="audio/surround/en/320kbit/segment_$Number$.m4s" initialization="audio/surround/en/320kbit/init.mp4" startNumber="1"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="4" mimeType="video/mp4" par="4096:1744" minBandwidth="258157" maxBandwidth="10285391" minWidth="422" maxWidth="4096" minHeight="180" maxHeight="1744" segmentAlignment="true" startWithSAP="1">
      <Representation id="240p 250kbps" frameRate="24" bandwidth="258157" codecs="avc1.4d400d" width="422" height="180">
        <SegmentTemplate timescale="1000" duration="2000" media="video/250kbit/segment_$Number$.m4s" initialization="video/250kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="480p 500kbps" frameRate="24" bandwidth="520929" codecs="avc1.4d4015" width="638" height="272">
        <SegmentTemplate timescale="1000" duration="2000" media="video/500kbit/segment_$Number$.m4s" initialization="video/500kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="480p 800kbps" frameRate="24" bandwidth="831270" codecs="avc1.4d4015" width="638" height="272">
        <SegmentTemplate timescale="1000" duration="2000" media="video/800kbit/segment_$Number$.m4s" initialization="video/800kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="576p 1100kbps" frameRate="24" bandwidth="1144430" codecs="avc1.4d401f" width="958" height="408">
        <SegmentTemplate timescale="1000" duration="2000" media="video/1100kbit/segment_$Number$.m4s" initialization="video/1100kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="720p 1500kbps" frameRate="24" bandwidth="1558322" codecs="avc1.4d401f" width="1277" height="544">
        <SegmentTemplate timescale="1000" duration="2000" media="video/1500kbit/segment_$Number$.m4s" initialization="video/1500kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="720p 2400kbps" frameRate="24" bandwidth="2487897" codecs="avc1.4d401f" width="1277" height="544">
        <SegmentTemplate timescale="1000" duration="2000" media="video/2400kbit/segment_$Number$.m4s" initialization="video/2400kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="1080p 3000kbps" frameRate="24" bandwidth="3113198" codecs="avc1.4d4028" width="1921" height="818">
        <SegmentTemplate timescale="1000" duration="2000" media="video/3000kbit/segment_$Number$.m4s" initialization="video/3000kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="1080p 4000kbps" frameRate="24" bandwidth="4149264" codecs="avc1.4d4028" width="1921" height="818">
        <SegmentTemplate timescale="1000" duration="2000" media="video/4000kbit/segment_$Number$.m4s" initialization="video/4000kbit/init.mp4" startNumber="1"/>
      </Representation>
      <Representation id="1080p 6000kbps" frameRate="24" bandwidth="6214307" codecs="avc1.4d4028" width="1921" height="818">
        <SegmentTemplate timescale="1000" duration="2000" media="video/6000kbit/segment_$Number$.m4s" initialization="video/6000kbit/init.mp4" startNumber="1"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="14" mimeType="text/vtt" lang="en">
      <Representation id="caption_en" bandwidth="256">
        <BaseURL>subtitles/subtitles_en.vtt</BaseURL>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="15" mimeType="text/vtt" lang="de">
      <Representation id="caption_de" bandwidth="256">
        <BaseURL>subtitles/subtitles_de.vtt</BaseURL>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="16" mimeType="text/vtt" lang="es">
      <Representation id="caption_es" bandwidth="256">
        <BaseURL>subtitles/subtitles_es.vtt</BaseURL>
      </Representation>
    </AdaptationSet>
    <AdaptationSet group="17" mimeType="text/vtt" lang="fr">
      <Representation id="caption_fr" bandwidth="256">
        <BaseURL>subtitles/subtitles_fr.vtt</BaseURL>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
```

On ExoPlayer 1.5.0 this stream has 9 available videotracks, while after upgrading it only has 5 available videotracks.

After a quick git-bisect it turns out that ExoPlayer started showing this behaviour after https://github.com/google/ExoPlayer/commit/f3113e7055bb2474d52f192a8fbacf752d35de22.

Apparently `MediaCodecInfo.areSizeAndRateSupported()` returns `false` for certain width and heights that played fine before: 

```
I/VideoFormatSelectorUtil: isSupported: true format.width: 422 format.height: 180 format.bitrate: 258157
I/VideoFormatSelectorUtil: isSupported: true format.width: 638 format.height: 272 format.bitrate: 520929
I/VideoFormatSelectorUtil: isSupported: true format.width: 638 format.height: 272 format.bitrate: 831270
I/VideoFormatSelectorUtil: isSupported: true format.width: 958 format.height: 408 format.bitrate: 1144430
I/VideoFormatSelectorUtil: isSupported: false format.width: 1277 format.height: 544 format.bitrate: 1558322
I/VideoFormatSelectorUtil: isSupported: false format.width: 1277 format.height: 544 format.bitrate: 2487897
I/VideoFormatSelectorUtil: isSupported: false format.width: 1921 format.height: 818 format.bitrate: 3113198
I/VideoFormatSelectorUtil: isSupported: false format.width: 1921 format.height: 818 format.bitrate: 4149264
```

I'm not sure whether this is intended behaviour or a bug, since these formats play fine under ExoPlayer 1.5.0-2.

**Comment by ojw28** [comment]
The device is telling the player that it's not capable of playing some of those streams, so we don't try and play them. I think this is working as intended as far as ExoPlayer is concerned.

This issue is most likely device specific. What device are you reproducing the issue on exactly, and on what version of Android? Could you paste the output of `adb shell getprop ro.build.fingerprint` on the affected device into this issue.

**Comment by RikHeijdens** [comment]
The device is a LG Nexus 5X, and the Android version is 6.0.1.

`rik@rheijdens:~$ adb shell getprop ro.build.fingerprint
google/bullhead/bullhead:6.0.1/MMB29K/2419427:user/release-keys`

I stepped into the MediaCodecInfo source with the debugger and it looks like that the MediaCodecInfo class decides that the device can not play the stream, however I am sure that the device is able to play the stream.

**Comment by ojw28** [comment]
It appears that by default, Android requires dimensions to be even rather than odd. See [MediaCodecInfo line 1080](https://android.googlesource.com/platform/frameworks/base/+/android-5.0.0_r2/media/java/android/media/MediaCodecInfo.java#1080) and the use of `mWidthAlignment` and `mHeightAlignment` in the [MediaCodecInfo.supports](https://android.googlesource.com/platform/frameworks/base/+/android-5.0.0_r2/media/java/android/media/MediaCodecInfo.java#975) method.

It's unclear whether the dimensions you're using are really supported, or just "happen to work". I'll follow up internally. Orthogonal to this, it feels like it would be preferable for you to use even dimensions when transcoding your content. I didn't read through in detail, but if you do an internet search for "h264 odd dimensions" you get a bunch of results suggesting that it's a bad idea.

**Comment by ojw28** [comment]
Also, is it possible for you to provide a sample stream (here or to `dev.exoplayer@gmail.com`). It would be interesting for us to take a look at what's in the stream itself.

**Comment by RikHeijdens** [comment]
I sent you an email with the stream.

I'm not completely sure if the mpd is wrong, e.g. one pixel off or that the actual content is transcoded in an odd width/height.

**Comment by ojw28** [comment]
The dimensions in the manifest do appear to be wrong. The actual widths are 1280 (where the manifest claims 1277) and 1920 (where the manifest claims 1921). The heights appear to be correct. So yes, correcting the manifest will fix this issue for you.

Supposing that your video streams really did have odd dimensions and that the manifest were correct, it's unclear whether we did the right thing in filtering. I suspect we did and that such dimensions really aren't supported, but I'll find out to make sure. Leaving this issue open for now to track this.
